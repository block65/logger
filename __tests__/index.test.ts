import {
  afterAll,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { createLoggerWithWaitableMock } from './helpers.js';

describe('Basic', () => {
  const oldEnv = process.env;
  const originalStackTraceLimit = Error.stackTraceLimit;

  beforeEach(() => {
    jest.resetModules();
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
    process.env = { ...oldEnv };
    Error.stackTraceLimit = originalStackTraceLimit;
  });

  afterAll(() => {
    process.env = oldEnv;
    Error.stackTraceLimit = originalStackTraceLimit;
  });

  test('Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: true });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object with undefined props', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: undefined });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('String, Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello', { omg: true });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('String Format, Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello %o', { omg: true });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object, String Format, String', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: true }, 'hello %s', 'world');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object, String Format, String', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();

    logger.warn({ omg: true }, 'hello %s:%s %d', 'world', 'and', 123);
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object + String = Format', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: true }, 'hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Auto (development)', async () => {
    process.env.NODE_ENV = 'development';
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Auto (production)', async () => {
    process.env.NODE_ENV = 'production';
    Error.stackTraceLimit = 10;
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Force true (production)', async () => {
    process.env.NODE_ENV = 'production';
    Error.stackTraceLimit = 10;
    const [logger, callback] = await createLoggerWithWaitableMock({
      traceCaller: true,
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Force false (development)', async () => {
    process.env.NODE_ENV = 'development';
    Error.stackTraceLimit = 10;
    const [logger, callback] = await createLoggerWithWaitableMock({
      traceCaller: false,
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Mixins basic + accum', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      mixins: [
        () => ({
          logger: {
            go: 'log',
          },
        }),
        () => ({
          luger: {
            go: 'bang',
          },
        }),
        (accum) => ({
          accum,
        }),
      ],
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Mixins property overwrite', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      mixins: [
        () => ({
          logger: {
            go: 'log',
          },
        }),
        () => ({
          logger: {
            go: 'brrr',
          },
        }),
      ],
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('allow undefined logger level', async () => {
    await createLoggerWithWaitableMock({
      level: undefined,
    });
  });
});
