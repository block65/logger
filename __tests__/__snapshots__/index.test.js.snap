// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Mixins basic + accum 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "logger": Object {
          "go": "log",
        },
        "luger": Object {
          "go": "bang",
        },
      },
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Mixins property overwrite 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "logger": Object {
          "go": "brrr",
        },
      },
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Object 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "omg": true,
      },
      "level": 40,
      "msg": undefined,
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Object with undefined props 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "omg": undefined,
      },
      "level": 40,
      "msg": undefined,
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Object, String = No Format 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "omg": true,
      },
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Object, String Format, ...Primitives 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "omg": true,
      },
      "level": 40,
      "msg": "hello world and 123",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Object, String Format, String 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "omg": true,
      },
      "level": 40,
      "msg": "hello world",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic String Format, Object 1`] = `
Array [
  Array [
    Object {
      "level": 40,
      "msg": "hello {\\"omg\\":true} spleen!",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic String, Object = Object ignored 1`] = `
Array [
  Array [
    Object {
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Trace Caller Auto (development) 1`] = `
Array [
  Array [
    Object {
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Trace Caller Auto (production) 1`] = `
Array [
  Array [
    Object {
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Trace Caller Force false (development) 1`] = `
Array [
  Array [
    Object {
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic Trace Caller Force true (production) 1`] = `
Array [
  Array [
    Object {
      "level": 40,
      "msg": "hello",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic allow flush and reuse 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "Before 1",
          "name": "Error",
          "stack": "Error: Before 1
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:172:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: Before 1],
      "level": 60,
      "msg": "Before 1",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "Before 2",
          "name": "Error",
          "stack": "Error: Before 2
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:173:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: Before 2],
      "level": 60,
      "msg": "Before 2",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "After 1",
          "name": "Error",
          "stack": "Error: After 1
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:175:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: After 1],
      "level": 60,
      "msg": "After 1",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "After 2",
          "name": "Error",
          "stack": "Error: After 2
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:176:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: After 2],
      "level": 60,
      "msg": "After 2",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic allow undefined logger level 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "Fake",
          "name": "Error",
          "stack": "Error: Fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:164:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: Fake],
      "level": 60,
      "msg": "Fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "Fake",
          "name": "Error",
          "stack": "Error: Fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:165:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: Fake],
      "level": 60,
      "msg": "Fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic context and release name co-exist 1`] = `
Array [
  Array [
    Object {
      "level": 10,
      "msg": "stuff",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "level": 20,
      "msg": "oh debug yes",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "level": 30,
      "msg": "Completed",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "Warn",
          "name": "Error",
          "stack": "Error: Warn
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:187:21)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: Warn],
      "level": 40,
      "msg": "Warn",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "Its an error",
          "name": "Error",
          "stack": "Error: Its an error
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:188:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: Its an error],
      "level": 50,
      "msg": "Its an error",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "OMG fatal",
          "name": "Error",
          "stack": "Error: OMG fatal
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:189:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
        },
      },
      "err": [Error: OMG fatal],
      "level": 60,
      "msg": "OMG fatal",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Basic log levels 1`] = `
Array [
  Object {
    "data": Object {
      "err": Object {
        "message": "OMG fatal",
        "name": "Error",
        "stack": "Error: OMG fatal
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/index.test.js:204:22)
    at Promise.then.completed (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:323:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/utils.js:249:10)
    at _callCircusTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:276:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:208:3)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:96:9)
    at _runTestsForDescribeBlock (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:90:9)
    at run (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/node_modules/jest-circus/build/run.js:31:3)",
      },
    },
    "err": [Error: OMG fatal],
    "level": 60,
    "msg": "OMG fatal",
    "time": 2009-02-13T23:31:30.000Z,
  },
]
`;
